ethcore/src/engines/clique/block_state.rs

```diff
@@ -0,0 +1,207 @@
+	pub next_timestamp_noturn: Option<SystemTime>, // noturn signing should wait until this time
+}
+
+impl CliqueBlockState {
+	/// Create new state with given information, this is used creating new state from Checkpoint block.
+	pub fn new(_author: Address, signers_sorted: Vec<Address>) -> Self {
+		return CliqueBlockState {
+			votes: Default::default(),
+			votes_history: Default::default(),
+			signers: signers_sorted,
+			recent_signers: Default::default(),
+			next_timestamp_inturn: None,
+			next_timestamp_noturn: None,
+		};
+	}
+
+	// see https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L474
+	fn verify(&self, header: &Header) -> Result<(Address), Error>{
+		let creator = recover_creator(header)?.clone();
+
+		// Check signer list
+		if !self.signers.contains(&creator) {
+			trace!(target: "engine", "current state: {:?}", self);
+			return Err(From::from(format!("Error applying #{}({}): {} is not in the signer list!",
+											header.number(),
+											header.hash(),
+											creator)));
+		}
+
+		// Check recent signer.
+		if self.recent_signers.contains(&creator) {
+			trace!(target: "engine", "current state: {:?}", self);
+			return Err(From::from(format!("Error applying #{}({}): {} is in the recent_signer list!",
+											header.number(),
+											header.hash(),
+											creator)));
+		}
+
+		// Ensure that the difficulty corresponds to the turn-ness of the signer
+		let inturn = self.is_inturn(header.number(), &creator);
+
+		if (inturn && *header.difficulty() != DIFF_INTURN) ||
+			(!inturn && *header.difficulty() != DIFF_NOTURN) {
+			return Err(From::from(format!("Error applying #{}({}): wrong difficulty!",
+											header.number(),
+											header.hash())));
+		}
+
+		Ok(creator)
+	}
+
+	/// Verify and apply an new header to current state, might fail with error.
+	pub fn apply(&mut self, header: &Header, is_checkpoint: bool) -> Result<Address, Error> {
+		let creator = self.verify(header)?;
+
+		// rotate recent signers.
+		self.recent_signers.push_front(creator.clone());
+		if self.recent_signers.len() >= ( self.signers.len() / 2 ) + 1 {
+			self.recent_signers.pop_back();
+		}
+
+		if is_checkpoint {
+			// checkpoint block should not affect previous tallying, so we check that.
+			let signers = extract_signers(header)?;
+			if self.signers != signers {
+				return Err(From::from("checkpoint block signers is different than expected"));
+			};
+
+			self.votes.clear();
+			self.votes_history.clear();
+
+			// maybe release some memory.
+			self.votes.shrink_to_fit();
+			self.votes_history.shrink_to_fit();
+
+			return Ok(creator);
+		}
+
+		let beneficiary = *header.author();
+
+		// No vote, ignore.
+		if beneficiary == NULL_AUTHOR {
+			return Ok(creator);
+		}
+
+		let nonce = *header.decode_seal::<Vec<&[u8]>>()?.get(1).ok_or(
+			"Error decoding seal"
+		)?;
+
+		let vote_type: VoteType;
```

Comment:
I don't like the style of `uninitialized values`, especially in such big functions!

Can you refactor this a bit and for example implement a method `fn from_slice(&[u8) -> Result<VoteType, Error>` for VoteType that takes care of checking the `nonce` slice?